File: password/python/password_validator.py
# password_validator.py

class PasswordValidator:
    def validate(self, password):
        missing_criteria = []

        if len(password) < 8:
            missing_criteria.append("at least 8 characters long")

        if not any(c.isupper() for c in password):
            missing_criteria.append("one uppercase letter")

        if not any(c.islower() for c in password):
            missing_criteria.append("one lowercase letter")

        if not any(c.isdigit() for c in password):
            missing_criteria.append("one number")

        special_characters = "!@#$%^&*()_+"
        if not any(c in special_characters for c in password):
            missing_criteria.append("one special character")

        if not missing_criteria:
            return "Password is valid."

        return "Password must contain " + ", ".join(missing_criteria) + "."

File: password/C#/PasswordValidator/PasswordValidator.cs
namespace PasswordValidator
{
    using System;
    using System.Collections.Generic;
    using System.Linq;

    public class PasswordValidator
    {
        public string Validate(string password)
        {
            var missingCriteria = new List<string>();

            if (password.Length < 8)
                missingCriteria.Add("at least 8 characters long");

            if (!password.Any(char.IsUpper))
                missingCriteria.Add("one uppercase letter");

            if (!password.Any(char.IsLower))
                missingCriteria.Add("one lowercase letter");

            if (!password.Any(char.IsDigit))
                missingCriteria.Add("one number");

            var specialCharacters = "!@#$%^&*()_+";
            if (!password.Any(c => specialCharacters.Contains(c)))
                missingCriteria.Add("one special character");

            if (missingCriteria.Count == 0)
                return "Password is valid.";

            return "Password must contain " + string.Join(", ", missingCriteria) + ".";
        }
    }
}
